core/src/pages.rs

```diff
@@ -0,0 +1,487 @@
+/// native page size, so can vary.
+pub const GEAR_PAGE_SIZE: usize = 0x4000;
+
+static_assertions::const_assert!(WASM_PAGE_SIZE < u32::MAX as usize);
+static_assertions::const_assert_eq!(WASM_PAGE_SIZE % GEAR_PAGE_SIZE, 0);
+
+/// Errors when act with PageU32Size.
+#[derive(Debug, Clone, derive_more::Display)]
+pub enum PageError {
+    /// Addition overflow.
+    #[display(fmt = "{_0} + {_1} overflows u32")]
+    AddOverflowU32(u32, u32),
+    /// Subtraction overflow.
+    #[display(fmt = "{_0} - {_1} overflows u32")]
+    SubOverflowU32(u32, u32),
+    /// Overflow U32 memory size: has bytes, which has offset bigger then u32::MAX.
+    #[display(fmt = "{_0} is too big to be number of page with size {_1}")]
+    OverflowU32MemorySize(u32, u32),
+    /// Cannot make pages iter from to.
+    #[display(fmt = "Cannot make pages iter from {_0} to {_1}")]
+    WrongRange(u32, u32),
+}
+
+/// Page number.
+#[derive(
+    Clone, Copy, Debug, Decode, Encode, PartialEq, Eq, PartialOrd, Ord, Hash, TypeInfo, Default,
+)]
+pub struct GearPage(pub(crate) u32);
+
+impl From<u16> for GearPage {
+    fn from(value: u16) -> Self {
+        // u16::MAX * GearPage::size() - 1 <= u32::MAX
+        static_assertions::const_assert!(GEAR_PAGE_SIZE <= 0x10000);
+        GearPage(value as u32)
+    }
+}
+
+impl From<GearPage> for u32 {
+    fn from(value: GearPage) -> Self {
+        value.0
+    }
+}
+
+/// Wasm page number.
+#[derive(Clone, Copy, Debug, Decode, Encode, PartialEq, Eq, PartialOrd, Ord, TypeInfo, Default)]
+pub struct WasmPage(pub(crate) u32);
+
+impl From<u16> for WasmPage {
+    fn from(value: u16) -> Self {
+        // u16::MAX * WasmPage::size() - 1 == u32::MAX
+        static_assertions::const_assert!(WASM_PAGE_SIZE == 0x10000);
+        WasmPage(value as u32)
+    }
+}
+
+impl From<WasmPage> for u32 {
+    fn from(value: WasmPage) -> Self {
+        value.0
+    }
+}
+
+/// Page number trait - page, which can return it number as u32.
+pub trait PageNumber: Into<u32> + Sized + Clone + Copy + PartialEq + Eq + PartialOrd + Ord {
+    /// Creates page from raw number.
+    ///
+    /// # Safety
+    /// This function is unsafe because it can create invalid page number.
+    unsafe fn from_raw(raw: u32) -> Self;
+
+    /// Returns raw (u32) page number.
+    fn raw(&self) -> u32 {
+        Into::<u32>::into(*self)
+    }
+
+    /// Checked subtraction.
+    fn checked_sub(&self, other: Self) -> Option<Self> {
+        self.raw()
+            .checked_sub(PageNumber::raw(&other))
+            .map(|p| unsafe { Self::from_raw(p) })
+    }
+
+    /// Returns iterator `self`..=`end`.
+    fn iter_end_inclusive(&self, end: Self) -> Result<PagesIterInclusive<Self>, PageError> {
+        if end.raw() >= self.raw() {
+            Ok(PagesIterInclusive {
+                page: Some(*self),
+                end,
+            })
+        } else {
+            Err(PageError::WrongRange(self.raw(), end.raw()))
+        }
+    }
+}
+
+impl PageNumber for WasmPage {
```

Comment:
I'd move these impls next to each corresponding types because all of structs and traits are in this file so we are not constrained by orphan rule anymore